docs: custom backend (#113)
explosion/spacy-llm: usage_examples/README.md

in the `spacy_llm/tasks/` directory. We also recommend checking out the
`spacy.NoOp.v1` task for a barebones implementation to pattern your task from.

## Writing your own backend

In `spacy-llm`, the [**backend**](../README.md#backend) is responsible for the

> interaction with the actual LLM model. For instance, this can be an

Reviewer: ```suggestion
interaction with the actual LLM model. The latter can be an
```

[API-based service](../README.md#spacyrestv1), or a local model - whether
you [downloaded it from the Hugging Face Hub](../README.md#spacydollyhfv1)
directly or finetuned it with proprietary data.

`spacy-llm` lets you implement your own custom backend so you can try out the

> latest LLM interface out there. It should be consistent with a given task since

Reviewer: "consistent with a given task" reads vague. Can you expand on this?

Reviewer: Yes, thanks! Just one remaining comment (see below).

behind the scenes, `spacy-llm` roughly performs the following pseudo-code:

```python
prompts = task.generate_prompts(docs)
responses = backend(prompts)

> task.parse_responses(docs, responses)

Reviewer: ```suggestion
docs = task.parse_responses(docs, responses)
```

```

Let's write a dummy backend that provides a random output for the
[text classification task](../README.md#spacytextcatv1).

```python
from spacy_llm.registry import registry
import random
from typing import Iterable

@registry.llm_backends("RandomClassification.v1")
def random_textcat(labels: str):
    labels = labels.split(",")
    def _classify(prompts: Iterable[str]) -> Iterable[str]:
        for _ in prompts:
            yield random.choice(labels)
    
    return _classify
```

```ini
...
[components.llm.task]
@llm_tasks = "spacy.TextCat.v1"
labels = LABEL1,LABEL2,LABEL3

[components.llm.backend]
@llm_backends = "RandomClassification.v1"
labels = ${components.llm.task.labels}  # Make sure to use the same label
...
```

Of course, this particular backend is not very realistic
(it does not even interact with an actual LLM model!).
But it does show how you would go about writing custom
and arbitrary logic to interact with any LLM implementation.

> Note that at the moment, all built-in tasks generate text-based prompts
> and expect text response. Therefore, built-in tasks and backends
> are currently inter-operable.

Reviewer: ```suggestion
Note that in all built-in tasks and backends prompts and responses are expected to be of type `string` 
- all built-in tasks and backends are therefore inter-operable. It's possible to work with arbitrary objects 
instead of `string` though - which might be useful if you want some third-party abstractions for prompts
or responses.
```

in the `spacy_llm/tasks/` directory. We also recommend checking out the
`spacy.NoOp.v1` task for a barebones implementation to pattern your task from.

## Writing your own backend

In `spacy-llm`, the [**backend**](../README.md#backend) is responsible for the
interaction with the actual LLM model. The latter can be an
[API-based service](../README.md#spacyrestv1), or a local model - whether
you [downloaded it from the Hugging Face Hub](../README.md#spacydollyhfv1)
directly or finetuned it with proprietary data.

`spacy-llm` lets you implement your own custom backend so you can try out the
latest LLM interface out there. Bear in mind that tasks are responsible for
creating the prompt and parsing the response – and both can be arbitrary objects.

> Hence, a backend should be consistent with the task you'd like it to run with.

Reviewer: ```suggestion
Hence, a backend's call signature should be consistent with that of the task you'd like it to run.
```
